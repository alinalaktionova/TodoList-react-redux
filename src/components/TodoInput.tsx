import React, {Dispatch} from "react";
import { connect } from "react-redux";
import {id} from "../actions/actionCreators";
import styled from "styled-components";
import {ActionType} from "../actions/ActionInterface";
import {ADD_TODO} from "../constants/ActionTypes";
import PropsInterface from "../interfaces/PropsInterface";

const TodoInputForm = styled.div`
  margin: 0 auto;
  width: 70%;
  display: flex;
  flex-direction: row;
`;

const InputField = styled.input`
  width: 85%;
  border-radius: 7px;
  overflow: auto;
  background-color: white;
  box-shadow: 0px 0.25rem 1rem rgba(black, 0.25);
`;

const StyledButton = styled.button`
  width: 15%;
  margin: 0 auto;
  padding: 5px;
  border-radius: 5px;
  background: dimgrey;
  color: white;
  font-size: 16px;
  text-transform: capitalize;
  letter-spacing: 1px;
  display: inline-block;
  cursor: pointer;
`;


let idEnc = id;

const TodoInput = (props: Partial<PropsInterface>) => {
  const inputRef = React.useRef<HTMLInputElement>(null);
  let [input, setInputState] = React.useState("");
  function handleInputChange(event: React.ChangeEvent<HTMLInputElement>) {
    setInputState(event.target.value);
  }

  return (
    <TodoInputForm>
      <InputField
        type="text"
        placeholder="add a todo item"
        onChange={event => handleInputChange(event)}
        ref={inputRef}
      />
      <StyledButton
        type="button"
        onClick={(event: React.MouseEvent) => {
          props.addTodo(input);
          if (inputRef && inputRef.current) {
            inputRef.current.value = "";
          }
        }}
      >
        + Add
      </StyledButton>
    </TodoInputForm>
  );
};

const mapDispatchToProps = (dispatch: Dispatch<ActionType>) => ({
   addTodo: (text: string) => dispatch({type: ADD_TODO, id: idEnc++, text}),
    dispatch
});
export default connect(null, mapDispatchToProps)(TodoInput);
