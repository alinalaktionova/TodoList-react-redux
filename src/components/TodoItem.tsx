import React, {Dispatch} from "react";
import {TOGGLE_COMPLETE_TODO, REMOVE_TODO} from "../constants/ActionTypes";
import { connect } from "react-redux";
import styled from "styled-components";
import PropsInterface from "../interfaces/PropsInterface";
import {ActionType} from "../actions/ActionInterface";


const Task = styled.li`
  align-items: center;
  display: flex;
  justify-content: space-between;
  list-style-type: none;
  position: relative;
  border: 1px solid #999
  border-radius: 3px
  display: block
  width: 100%
  box-sizing: border-box
  outline: none;
  font-size: 16px;
  margin-bottom: 20px;
`;

const DeleteBtn = styled.span`
  display: block;
  position: absolute;
  top: 12px;
  left: 10px;
  font-size: 20px;
  color: dimgrey;
  cursor: pointer;
`;


const TodoItem = (props: Partial<PropsInterface>) => {
  console.dir(props);
  return (
    <Task
      style={{
        textDecoration: props.todo.isCompleted ? "line-through" : "none"
      }}
      onClick={() => props.toggleTodoComplete(props.todo.id)}
    >
      <p>{props.todo.text}</p>
      <DeleteBtn>
        <i
          className="fas fa-trash-alt"
          onClick={() => props.removeTodo(props.todo.id)}
        />
      </DeleteBtn>
    </Task>
  );
};

const mapDispatchToProps = (dispatch: Dispatch<ActionType>) => ({
  toggleTodoComplete: (id: number) => dispatch({type: TOGGLE_COMPLETE_TODO, id: id}),
  removeTodo: (id: number) => dispatch({type: REMOVE_TODO, id: id }),
  dispatch
});

export default connect(null, mapDispatchToProps)(TodoItem);
