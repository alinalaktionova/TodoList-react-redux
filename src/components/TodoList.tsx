import React, { Dispatch } from "react";
import TodoItem from "./TodoItem";
import { connect } from "react-redux";
import {DELETE_ALL} from '../constants/ActionTypes'
import TodoInterface from "../interfaces/TodoInterface";
import styled from "styled-components";
import PropsInterface from "../interfaces/PropsInterface";
import {ActionType} from "../actions/ActionInterface";

const TaskList = styled.ul`
  display: flex;
  flex-direction: column;
  justify-content: center;
  width: 70%;
  text-align: center;
  margin: 30px auto;
  padding: 0;
`;

const ClearButton = styled.button`
  width: 20%;
  margin: 0 auto;
  padding: 5px;
  border-radius: 5px;
  background: dimgrey;
  color: white;
  font-size: 16px;
  text-transform: capitalize;
  letter-spacing: 1px;
  display: inline-block;
  cursor: pointer;
`;


const TodoList = (props: Partial<PropsInterface>) => {
  return (
    <TaskList>
      {props.todos.map((todo: TodoInterface) => {
        return (
          <TodoItem
            key={todo.id}
            todo={todo}
          />
        );
      })}
      <ClearButton type="button" onClick={(event: React.MouseEvent) => props.deleteAll()}>
        delete all
      </ClearButton>
    </TaskList>
  );
};


const mapDispatchToProps = (dispatch: Dispatch<ActionType>) => ({
  deleteAll: () => dispatch({type: DELETE_ALL}),
  dispatch
});

export default connect(null, mapDispatchToProps)(TodoList);
