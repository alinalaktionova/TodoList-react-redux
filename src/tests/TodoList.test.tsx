import React from "react";
import TodoList from "../components/TodoList";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import ReactTestUtils from "react-dom/test-utils";
import configureStore, { MockStore } from "redux-mock-store";
import TodoItem from "../components/TodoItem";
import {
  addTodo,
  deleteAll,
  removeTodo,
  toggleTodoComplete
} from "../actions/actionCreators";
import TodoInput from "../components/TodoInput";
import TodoInterface from "../interfaces/TodoInterface";

const mockStore = configureStore([]);

describe("My Connected React-Redux Component", () => {
  let store: MockStore;
  let component: any;
  let todos: TodoInterface[];
  beforeEach(() => {
    store = mockStore({ todos: [] });
    store.dispatch = jest.fn();
    todos = [{...store.getState()}];
    component = renderer.create(
      <Provider store={store}>
        <TodoList todos={todos}/>
      </Provider>
    );
  });
  it("should render with given state from Redux store", () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should dispatch an action on button click", () => {
    renderer.act(() => {
      component.root.findByType("button");
    });
    ReactTestUtils.Simulate.click(component, store.dispatch(deleteAll()));
    expect(component.toJSON()).toMatchSnapshot();
  });
});
