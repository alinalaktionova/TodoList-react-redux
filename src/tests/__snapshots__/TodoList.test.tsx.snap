// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Connected React-Redux Component should dispatch an action on button click 1`] = `
<ul
  className="sc-htpNat lcxLoP"
>
  <li
    className="sc-bdVaJa fbSIgk"
    onClick={[Function]}
    style={
      Object {
        "textDecoration": "none",
      }
    }
  >
    <p />
    <span
      className="sc-bwzfXH ibHVFN"
    >
      <i
        className="fas fa-trash-alt"
        onClick={[Function]}
      />
    </span>
  </li>
  <button
    className="sc-bxivhb hAhGGw"
    onClick={[Function]}
    type="button"
  >
    delete all
  </button>
</ul>
`;

exports[`My Connected React-Redux Component should render with given state from Redux store 1`] = `
<ul
  className="sc-htpNat lcxLoP"
>
  <li
    className="sc-bdVaJa fbSIgk"
    onClick={[Function]}
    style={
      Object {
        "textDecoration": "none",
      }
    }
  >
    <p />
    <span
      className="sc-bwzfXH ibHVFN"
    >
      <i
        className="fas fa-trash-alt"
        onClick={[Function]}
      />
    </span>
  </li>
  <button
    className="sc-bxivhb hAhGGw"
    onClick={[Function]}
    type="button"
  >
    delete all
  </button>
</ul>
`;
